src/storage/table/struct_column_data.cpp

```diff
@@ -124,19 +126,37 @@ static Vector &GetFieldVectorForScan(Vector &result, optional_idx field_index) {
 	return *children[index];
 }
 
+static bool CastIsPushedDown(ColumnData &child, ColumnScanState &child_state) {
+	if (!child_state.storage_index.HasType()) {
+		return false;
+	}
+	auto &scan_type = child_state.storage_index.GetType();
+	return child.GetType() != scan_type;
+}
+
 idx_t StructColumnData::Scan(TransactionData transaction, idx_t vector_index, ColumnScanState &state, Vector &result,
                              idx_t target_count) {
 	auto scan_count = validity->Scan(transaction, vector_index, state.child_states[0], result, target_count);
 	IterateFields(
-	    state, [&](ColumnData &field, optional_idx field_vector_index, ColumnScanState &field_state, bool should_scan) {
+	    state, [&](idx_t child_index, optional_idx field_vector_index, ColumnScanState &field_state, bool should_scan) {
 		    auto &target_vector = GetFieldVectorForScan(result, field_vector_index);
 		    if (!should_scan) {
 			    // if we are not scanning this vector - set it to NULL
 			    target_vector.SetVectorType(VectorType::CONSTANT_VECTOR);
 			    ConstantVector::SetNull(target_vector, true);
 			    return;
 		    }
-		    field.Scan(transaction, vector_index, field_state, target_vector, target_count);
+		    auto &field = *sub_columns[child_index];
+		    auto &child_type = StructType::GetChildTypes(type)[child_index].second;
+		    if (CastIsPushedDown(field, field_state)) {
+			    D_ASSERT(state.storage_index.IsPushdownExtract());
+			    Vector intermediate(child_type, target_count);
+			    field.Scan(transaction, vector_index, field_state, intermediate, target_count);
+			    //! FIXME: this is an optional ptr.. when is it not set??
+			    VectorOperations::Cast(*state.context.GetClientContext(), intermediate, target_vector, target_count);
```

Comment:
Maybe this should be a generic expression instead of a cast - where we keep the cast state in the `ColumnScanState`. This could prevent re-initializing and such, and would be more flexible in the future.

Reply:
For example, one optimization this could enable in the future is pushing expressions into the projection which is useful if they exist in both the filter and the projection, e.g.:

```sql
SELECT regexp_extract('...') AS extracted
FROM tbl
WHERE extracted = ...
```

Currently we need to execute `regexp_extract` twice (once in the filter, once in the projection) - but allowing expressions to be pushed would avoid this.